### DISCUSSION TITLE
Understanding Cache Action keys

### DISCUSSION THREAD

**Discussion by DannyBen** [community/community#25027] [Actions]
<p>Can someone please help me understand how the cache keys work in the Cache Action?</p>
<p>Here are the things I do not fully understand.</p>
<ol>
<li>
<p>In some places, I see people using the “github.sha” as part of the cache key. For example:</p>
<p>key: ${{ runner.os }}-docker-${{ github.sha }}<br>
restore-keys: |<br>
${{ runner.os }}-docker-</p>
</li>
</ol>
<p>So, as I understand it, the cache key will change with every commit, and therefore, not be a “cache hit”. What reason is there to use the commit sha in the cache key?</p>
<ol start="2">
<li>What are restore keys</li>
</ol>
<p>The <a href="https://github.com/actions/cache#inputs" rel="nofollow noopener">documentation on restore-keys</a> states that there are “fallback keys” if no cache hit occurred. So does it mean that when caching is done, it will be cached using two keys, and the topmost key will be selected on restore? Or, is the “restore-keys” option actually listing <strong>prefixes</strong> of keys, so its meaning is more like “${{ runner.os }}-docker- <strong>*</strong>”?</p>
<p>If the answer is the latter, then it explains the first question as well, at least to some extent.</p>
<p>If these are not prefixes, then why is everybody using a trailing hyphen in the restore key names?</p>
<ol start="3">
<li>Restore key documentation</li>
</ol>
<p>A quote from the restore key documentaiton:</p>
<ul>
<li>restore-keys - An ordered list of keys to use for restoring the cache if no cache hit occurred for key</li>
</ul>
<p>But the syntax everywhere, is actually used like this:</p>
<pre><code>restore-keys: |
  ${{ runner.os }}-docker-
</code></pre>
<p>which is not an ordered list, but rather a newline delimited string. I expected the syntax to match the definition (and actually be an oordered list), which makes more sense to me here:</p>
<pre><code>restore-keys:
  - ${{ runner.os }}-docker-
  - ${{ runner.os }}-
</code></pre>
<p>What is the reason for this variable typed as a newline delimited string, instead of a proper YAML array?</p>
<p>Sorry for the long post, hoping someone has the patience to elaborate, and if this is you - thanks in advance!</p>

**Comment by BrightRan** [comment, marked as answer]
<p>Caches not only can be shared between workflow runs, you also can share caches between jobs in the same workflow run. So when using <strong>github.sha</strong> in the key is feasible. The key can be any combination of variables, context values, static strings, and functions.<br>
An ordered list of alternative keys to use for finding the cache if no cache hit occurred for key. If you provide restore-keys, the cache action sequentially searches for any caches that match the list of restore-keys.</p>
<ol>
<li>When there is an exact match, the action restores the files in the cache to the path directory.</li>
<li>If there are no exact matches, the action searches for partial matches of the restore keys. When the action finds a partial match, the most recent cache is restored to the path directory.</li>
</ol>
<p>More details, you can reference the documentation about <a href="http://help.github.com/en/actions/automating-your-workflow-with-github-actions/caching-dependencies-to-speed-up-workflows#example-using-the-cache-action" rel="nofollow noopener">Caching dependencies to speed up workflows</a>.</p>
